server/api.go

```diff
@@ -223,6 +223,40 @@ func (d dexAPI) ListPasswords(ctx context.Context, req *api.ListPasswordReq) (*a
 
 }
 
+func (d dexAPI) VerifyPassword(ctx context.Context, req *api.VerifyPasswordReq) (*api.VerifyPasswordResp, error) {
+	if req.Email == "" {
+		return nil, errors.New("no email supplied")
+	}
+
+	if req.Password == "" {
+		return nil, errors.New("no passwored to verify supplied")
+	}
+
+	// TODO we can dry this up if https://github.com/coreos/dex/pull/1271/files merges
+	password, err := d.s.GetPassword(req.Email)
+	if err != nil {
+		if err == storage.ErrNotFound {
+			return &api.VerifyPasswordResp{
+				NotFound: true,
+				Verified: false,
+			}, nil
+		}
+		d.logger.Errorf("api: there was an error retrieving the password: %v", err)
+		return nil, fmt.Errorf("verify password: %v", err)
+	}
+
+	resp := &api.VerifyPasswordResp{
+		NotFound: false,
+	}
+
+	if err := bcrypt.CompareHashAndPassword(password.Hash, []byte(req.Password)); err != nil {
```

Comment:
What do you think, would it make sense to inspect the `err` return here? From the code, it'll return [`bcrypt.ErrMismatchedHashAndPassword`](https://github.com/golang/crypto/blob/80fca2ff14a3fb7db622c9a5462f8ce9ea026cba/bcrypt/bcrypt.go#L115) if the passwords don't match; but there's other error returns that indicate different failures: [here](https://github.com/golang/crypto/blob/80fca2ff14a3fb7db622c9a5462f8ce9ea026cba/bcrypt/bcrypt.go#L101-L103) and [here](https://github.com/golang/crypto/blob/80fca2ff14a3fb7db622c9a5462f8ce9ea026cba/bcrypt/bcrypt.go#L106-L108).

I'd propose something along the lines of
```go
if err := bcrypt.CompareHashAndPassword(password.Hash, []byte(req.Password)); err != nil {
	if err == bcrypt.ErrMismatchedHashAndPassword {
		resp.Verified = false
		return resp, nil
	}
	return nil, err
}
```

There might be some further simplifications around those three returns, so, just illustrating what I had in mind.